hail/src/main/scala/is/hail/expr/ir/functions/LocusFunctions.scala

```diff
@@ -172,12 +171,12 @@ object LocusFunctions extends RegistryFunctions {
     val locusClass = Locus.getClass
 
     registerCode("contig", tlocus("T"), TString,
-      (x: PType) => -x.asInstanceOf[PLocus].contigType) {
+      (x: PType) => x.asInstanceOf[PLocus].contigType) {
       case (r, rt, (locusT: PLocus, locus: Code[Long])) =>
         locusT.contig(locus)
     }
 
-    registerCode("position", tlocus("T"), TInt32, null) {
+    registerCode("position", tlocus("T"), TInt32, (x: PType) => x.asInstanceOf[PLocus].positionType) {
```

Comment:
Assert that the position type is required. Otherwise, this could break.

Reply:
That's subtle. Would you be open to moving (in a followup PR), this assertion to returnPType. The proposal there is that we do not perform setRequired if pt is provided. Rationale: our "calculate ptype" anonymous function doesn't fully calculate its ptype, so the user of this function needs to always keep in mind that they don't have full control over the returned outer type (but do have control on any nested types). This can lead to subtle bugs like this. Another subtle example is: in master currently we have for "contig" the return ptype as `-x.asInstanceOf[PLocus].contigType`. This doesn't do anything, since returnPType overrides this requiredeness.

Reply:
I thought about this a little when I got up this morning -- Since this is `registerCode`, the missingness of the result only depends on the missingness of the inputs. There's no way it can be missing unless the input locus is missing. If we add support for loci with missing positions/contigs (what does that even mean?) then the code generated will need to be changed too.

Basically, I was wrong, this is fine.

Reply:
Can you explain why the assertion is needed actually? In the function body we call `locusT.position(locus)`. Function signature: `def position: (off: Code[Long]): Code[Int]`, and in the canonical implementation  Region.loadInt(representation.loadField(off)) is called without a null pointer check in loadField (or a missingness check...no possibility of returning null from these. functions). So the return ptype should always be required, which returnPType forces (although again, I think we should make returnPType not modify the ptype returned from the "calculate ptype" anonymous function).

Reply:
> , I think we should make returnPType not modify the ptype returned from the "calculate ptype" anonymous function).

Can you explain this part? We do need the `setRequired(args.forall(_.reqiured))` here.

Reply:
I was thinking of:

```scala
override def returnPType(argTypes: Seq[PType], returnType: Type): PType = {
  if(pt == null) {
    p = PType.canonical(returnType)
    return p.setRequired(argTypes.forall(_.required))
  }

  assert(p.required >= argTypes.forall(_.required))
  p
}
```

If I choose to override the default, I (the caller) should do the right thing based on my understanding of the body that calculates the value.

Reply:
While returnPType as it stands saves typing, it means I do weird things like returning a non-required PInt32() from the function that is supposed to be inferring the ptype, only because I know that somewhere else this return ptype will be modified. I think this adds cognitive overhead.

Reply:
> If I choose to override the default, I (the caller) should do the right thing based on my understanding of the body that calculates the value.

This means every single registerCode would have to do argument requiredness unification, right? That seems quite error-prone.

>  it means I do weird things like returning a non-required PInt32() from the function that is supposed to be inferring the ptype

I feel like the right pattern is to return a required value, and then set it to optional if any of the arguments is optional. That doesn't strike me as particularly indirect.